LiS/man/man9/fifo.9.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: fifo.9.man,v 1.1.4.4 2008-04-29 08:33:21 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:21 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: fifo.9.man,v $
.\" Revision 1.1.4.4  2008-04-29 08:33:21  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.3  2006/04/05 09:42:43  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.TH FIFO 9 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\" =========================================================================
.\"
.\" Copyright (C) 2000  John A. Boyd Jr.  protologos, LLC
.\"
.\" This library is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Library General Public
.\" License as published by the Free Software Foundation; either
.\" version 2 of the License, or (at your option) any later version.
.\"
.\" This library is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public
.\" License along with this library; if not, write to the
.\" Free Software Foundation, Inc., 59 Temple Place - Suite 330, Cambridge,
.\" MA 02139, USA.
.\"
.\" =========================================================================
.SH NAME
.B fifo
\- pseudo-driver for
.BR STREAMS "\-based"
.I fifos
.SH DESCRIPTION
The
.B fifo
pseudo-driver
(which is internal to LiS) provides
.BR STREAMS "\-based" 
.I fifos
as single character special files, and
.BR STREAMS "\-based" 
.I pipes
as pairs of character special files which are interconnected (see
.BR pipe "(3)).  "
.sp
.BR STREAMS "\-based"
.I fifos
differ from typical
.BR STREAMS "\-based"
character special files in that there are not separate
.I "stream head"
and
.I driver
queue pair within the
.BR STREAMS "\-based"
file.  Instead, a
.I fifo
is created with only a single queue pair for the
.IR "stream head" .
Moreover, in a typical
.I driver
queue pair, the write queue is not connected to a next queue.
In a
.IR fifo ,
the write queue is directed to the read queue of the pair.
A
.I pipe
comprises a pair of
.IR fifos ", "
with the write queue of each pair directed to the read queue of the other.
The two
.B fifos
comprising a
.B pipe
are referred to as
.IR peers ", "
and each somewhat represents a
.I driver
to the other.
As a degenerate case, a
.B fifo
is its own peer.
.sp
.B STREAMS
.I modules
may be pushed onto
.B fifos
and
.BR pipes ", "
but should not expect a
.I driver
below them; instead, the
.BR SAMESTR "()"
function should be used from the write queue of a pair to determine if
the module is the lowest in the
.BR STREAMS "\-based"
file (this is called the
.IR midpoint "). "
The structure of a
.B fifo
or
.B pipe
is preserved when modules are pushed (and popped); i.e., the write
queue at the
.I midpoint
will always be directed at the read queue of the
.IR peer "."
.sp
Input and output are handled at a
.B fifo
stream head as they would normally be handled
at a stream head.  In LiS, an
.BR "fifo open" "() "
entry point exists to assign minor device numbers to new opens under the
.B fifo
major device number, and a
.BR close "() "
entry point is used correspondingly to release them.  These functions
are kept in a
.I streamtab
data structure (as they would normally be for any
.B STREAMS
driver or module) which is private to the LiS implementation.
.SH "APPLICATION USAGE"
In the current Linux kernels, character special major numbers are limited
to 16 bits, and major and minor device numbers to 8 bis each.  This limits
a system to 256 total major device numbers and 256 total minor devices
per major device number.  This is a rather severe limitation where
mechanisms like
.B fifos
and
.B pipes
are concerned.
.sp
However, a driver may handle more than one major device number.  The
.B fifo
pseudo-driver uses this to overcome this limitation, by supporting the
automatic allocation and use of multiple major device numbers for
.B fifos
and
.BR pipes .
Specifying more than 256 minor devices is done in the usual manner, i.e.,
by specifying the number of "units" in the appropriate
.B Config
file.  Enough major device numbers will be allocated to cover the
requested number of minor devices (if available, else an error will
occur in 
.BR strconf "(8)). "
The number allocated will include one minor device per major number
to be used as a
.BR fifo "\-specific"
.I clone
minor device (specifically, minor number 0), which exhibits special behavior.
.sp
Normally, when cloning is done via the
.B clone
pseudo-driver, the
.B clone
major device number is used, along with the desired actual major number
as the minor device number.  When an
.BR open "() "
is performed on such a device, the
.BR "clone open" "() "
routine in turn calls the appropriate driver's
.BR open "(), "
with the
.I sflag
parameter set to
.BR CLONEOPEN .
The driver's
.BR open "() "
is expected in this case to allocate an unused minor device number,
and return it via an entirely new device number in the
.I "devp" 
parameter.  In this way, a driver can change the device number to be
used for a
.BR STREAMS "\-based"
file.
When minor device 0 for a specified for a
.B fifo
major device, the driver will also clone a new minor device number.
However, LiS opens
.B fifo
devices differently; specifically, when an already-opened
.BR fifo "\-specific"
clone minor device is reopened, the new and subsequent opens will use
the already-opened clone.
Thus, using minor device 0 for a
.B fifo
when creating a file system node will ensure that all concurrent
opens of the associated path name will use the same
.BR STREAMS "\-based"
file; at the same time, opens of different file system nodes via
different paths will open their respectively different
.BR STREAMS "\-based"
files.
This is essentially how kernel-based
.B fifos
behave -applications and users of
.BR STREAMS "\-based"
.B fifos
don't have to keep track of minor numbers to achieve this same behavior
when it is desired.
.sp
It is in fact recommended that only two forms of file system nodes be used for
.BR STREAMS "\-based " fifos ": "
the
.B clone
major number as major number with a
.B fifo
major number as minor number, to be used when
.I every
open of the associated path must clone a new
.BR fifo ", "
and a
.B fifo
major number as major number with 0 as the minor number, to be used
when new opens are to clone a new
.B fifo
but subsequent concurrent opens are to use the already opened
.BR fifo ". "
These are represented by two device special file paths created when LiS
is installed:
.RB "/" dev "/" fifo
for the former, and
.RB "/" dev "/" fifo.0
for the latter.  It is recommended that these be used, possibly
along with the equivalent of
.BR stat "(2) "
to determine appropriate major device numbers for the
.B clone
and
.B fifo
pseudo-drivers, which are also determined when LiS is installed.
It can be noted that
.B pipes
are actually created as instances of the former, after which the write
queues are peer-connected.
.sp
The
.B fifo
pseudo-driver allocates minor devices in
.I round-robin
fashion; i.e., a list of available minor devices is kept, and once a
minor number is finally closed, it is put at the end of this list.
Thus, a
.B fifo
minor device which is opened and closed will not be immediately reused.
.SH WARNINGS
Because
.BR STREAMS "\-based"
.B fifos
and
.B pipes
are implemented as character special devices, they do not appear as
.B pipe
devices when examined with
.BR stat "(2) "
or the equivalent (e.g., 
.BR ls "(1)); "
i.e. the
.B S_IFIFO
indication is not set in the mode -
.B S_IFCHR
is set instead, and the actual device number is indicated in the
.I st_rdev
field of the
.I stat
data structure.
.sp
Because of the potential use of multiple major numbers, applications
should not depend on a
.B fifo
or
.B pipe
having a specific major device number, nor should an application depend on
all
.B fifos
and
.B pipes
having the same major device number.
.SH "SEE ALSO"
.BR clone "(9), "
.BR connld "(9), "
.BR fifo "(4), "
.BR ls "(1), "
.BR pipe "(3), "
.BR pipemod "(9), "
.BR STREAMS "(4), "
.BR stat "(2), "
.BR strconf "(8)"
.SH HISTORY
Unix System V Release 4 (SVR4)
.SH BUGS
Please report bugs (and/or comments and suggestions) to the LiS mailing list.
.SH AUTHOR
John Boyd, protologos LLC.


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/fifo.c
/*****************************************************************************

 @(#) $RCSfile: fifo.c,v $ $Name:  $($Revision: 1.1.1.2.4.14 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: fifo.c,v $
 Revision 1.1.1.2.4.14  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 1.1.1.2.4.13  2007/08/15 04:57:58  brian
 - GPLv3 updates

 *****************************************************************************/

#ident "@(#) $RCSfile: fifo.c,v $ $Name:  $($Revision: 1.1.1.2.4.14 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] = "$RCSfile: fifo.c,v $ $Name:  $($Revision: 1.1.1.2.4.14 $) $Date: 2008-04-29 08:33:11 $";

/*
 *  fifo.c - FIFO pseudo-driver
 *
 *  This pseudo-driver is used for FIFOs and pipes.  Note that the queue
 *  pairs for this pseudo-driver become the stream head; thus the use of
 *  strrput and strwsrv in the qinit structures.
 *
 *  Since Linux has FIFOs and pipes implemented in the kernel (and they
 *  monopolize the S_IFIFO file type), this pseudo-driver manages FIFOs
 *  as character minor devices.  Since Linux also limits the number of
 *  minor devices per major number to 256, this pseudo-driver manages
 *  multiple major numbers in order to support more than 256 minor devices.
 *  They are list-managed, and so can be allocated/freed in effectively
 *  constant time.  The open()/close() entry points in this file serve
 *  primarily to support the FIFO minor device pool; the remainder of
 *  the necessary support for FIFOs and pipes is in the stream head code.
 *
 *  Minor device 0 for each major is a clonable; it will never be actually
 *  used as an open stream.  Thus, 255 minor numbers are available per
 *  major number.
 *
 *  Copyright (C) 2000  John A. Boyd Jr.  protologos, LLC
 */

#include <sys/LiS/module.h>	/* should be first */

#include <sys/LiS/config.h>

#include <sys/stream.h>
#include <sys/stropts.h>
#ifdef LINUX
#include <linux/errno.h>
#else
#include <sys/errno.h>
#endif
#include <sys/cmn_err.h>

#include <sys/LiS/head.h>	/* for lis_strrput/lis_strwsrv */
#include <sys/lislocks.h>	/* semaphores */

#include <linux/list.h>
#include <sys/osif.h>

/*
 *  Some configuration sanity checks
 */
#ifndef FIFO__DRV
#error Not configured
#endif

#ifndef FIFO__ID
#define FIFO__ID 0xfef0
#endif

#ifndef FIFO__CMAJORS
#error The number of major numbers must be defined
#endif

#ifndef FIFO__UNITS
#define FIFO__UNITS 256
#endif

#define FIFO_MAX_MINOR ((FIFO__UNITS/FIFO__CMAJORS) - 1)

#define  MOD_ID   FIFO__ID
#define  MOD_NAME "fifo"

/*
 *  per-minor device (i.e., per-open) info
 */
typedef struct _fifo_dev {
	short index;			/* self-index */
	short flags;			/* see below */
	struct list_head list;		/* for free/used lists */
} fifo_dev_t;

#define FIFO_CLONABLE    0x01
#define FIFO_ALLOCATED   0x02

int fifo_units = FIFO__UNITS;		/* total number of units (i.e., minors) */

fifo_dev_t fifo_dev[FIFO__UNITS];	/* elems 0 mod minor() not used */

static LIST_HEAD(free_fifos);
static int num_free_fifos = 0;

#if 0
static LIST_HEAD(used_fifos);
static int num_used_fifos = 0;
#endif

extern int fifo__0_majors[];		/* in streams.o via modconf */

static lis_semaphore_t fifo_sem;	/* for allocating minor numbers */

static int fifo_initialized = 0;

/*
 *  function prototypes
 */
static int _RP fifo_open(queue_t *, dev_t *, int, int, cred_t *);
static int _RP fifo_close(queue_t *, int, cred_t *);

/*
 *  module structure
 */
static struct module_info fifo_minfo = {
	MOD_ID,				/* id */
	MOD_NAME,			/* name */
	0,				/* min packet size accepted */
	INFPSZ,				/* max packet size accepted */
	10240L,				/* highwater mark */
	512L				/* lowwater mark */
};

static struct qinit fifo_rinit = {
	lis_strrput,			/* put */
	lis_strrsrv,			/* service */
	fifo_open,			/* open */
	fifo_close,			/* close */
	NULL,				/* admin */
	&fifo_minfo,			/* info */
	NULL				/* stat */
};

static struct qinit fifo_winit = {
	NULL,				/* put */
	lis_strwsrv,			/* service */
	NULL,				/* open */
	NULL,				/* close */
	NULL,				/* admin */
	&fifo_minfo,			/* info */
	NULL				/* stat */
};

struct streamtab fifo_info = {
	&fifo_rinit,			/* read queue init */
	&fifo_winit,			/* write queue init */
	NULL,				/* lower mux read queue init */
	NULL				/* lower mux write queue init */
};

static inline int
fifo_major_index(int maj)
{
	int idx;

	for (idx = 0; idx < FIFO__CMAJORS; idx++)
		if (fifo__0_majors[idx] == maj)
			return (idx);

	return (-ENOENT);
}

static inline int
fifo_unit_index(dev_t dev)
{
	int idx = fifo_major_index(getmajor(dev));
	int maxminors = FIFO_MAX_MINOR + 1;

	if (idx < 0)
		return (idx);

	idx = (idx * maxminors) + getminor(dev);

	if (idx >= fifo_units)
		return (-ENXIO);

	return (idx);
}

static dev_t
fifo_mkdev(int unit_index)
{
	int maxminors = FIFO_MAX_MINOR + 1;

	return (makedevice(fifo__0_majors[(unit_index / maxminors)], unit_index % maxminors));
}

static int
fifo_alloc(int idx)
{
	fifo_dev_t *fdp;

	if (idx > fifo_units)
		return -ENXIO;
	if (list_empty(&free_fifos))
		return -EMFILE;

	if (idx < 0) {
		struct list_head *e = free_fifos.next;

		fdp = list_entry(e, fifo_dev_t, list);
	} else
		fdp = &(fifo_dev[idx]);

	if (fdp->flags & FIFO_ALLOCATED)
		return -EBUSY;

	fdp->flags |= FIFO_ALLOCATED;

	list_del(&(fdp->list));
	num_free_fifos--;
#if 0
	list_add(&(fdp->list), &(used_fifos));
	num_used_fifos++;
#endif

	return fdp->index;
}

static void
fifo_free(int idx)
{
	fifo_dev_t *fdp;

	if (idx < 0 || idx >= fifo_units)
		return;

	fdp = &(fifo_dev[idx]);

	if (!(fdp->flags & FIFO_ALLOCATED))
		return;

	fdp->flags &= ~FIFO_ALLOCATED;
#if 0
	list_del(&(fdp->list));
	num_used_fifos--;
#endif
	list_add(&(fdp->list), free_fifos.prev);
	num_free_fifos++;
}

/*
 *  open
 */
static int _RP
fifo_open(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *credp)
{
	int m;

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_open( 0x%p, 0x%x, 0x%x, 0x%x, ... )\n", MOD_NAME, q, *devp, flag,
		sflag);
#endif

	lis_down(&fifo_sem);

	/* 
	 *  we allow a clone open both ways: 0 minors are clonables
	 */
	m = (sflag == CLONEOPEN ? 0 : fifo_unit_index(*devp));

	if (m < 0) {
		lis_up(&fifo_sem);
		return (m);	/* m is an error code */
	}

	if (fifo_dev[m].flags & FIFO_CLONABLE) {
		/* 
		 *  select any unused unit
		 */
		if ((m = fifo_alloc(-1)) < 0) {
			lis_up(&fifo_sem);
			return (m);	/* m is an error code */
		}
	} else {
		/* 
		 *  select the requested unit if unused
		 */
		if (!(q->q_ptr)) {
			if ((m = fifo_alloc(m)) < 0) {
				lis_up(&fifo_sem);
				return (m);
			}
		}
	}

	if (!q->q_ptr) {
		RD(q)->q_ptr = WR(q)->q_ptr = (void *) &(fifo_dev[m]);

		/* 
		 *  simulate a FIFO by linking WR(q) -> RD(q);  the head might
		 *  also do this, but just in case...
		 */
		WR(q)->q_next = RD(q);

		qprocson(q);

		MODGET();
	}

	*devp = fifo_mkdev(m);

	lis_up(&fifo_sem);

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_open( 0x%p, 0x%x, ... ) unit %d OK\n", MOD_NAME, q, *devp, m);
#endif

	return 0;		/* OK */
}

/*
 *  close
 *
 *  Pipe linkage is disassembled by lis_qdetach.
 */
static int _RP
fifo_close(queue_t *q, int flag, cred_t *credp)
{
	fifo_dev_t *fdp = (fifo_dev_t *) q->q_ptr;

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_close( 0x%p, 0x%x, ... ) unit %d\n", MOD_NAME, q, flag,
		fdp ? fdp->index : 0);
#endif

	if (fdp == NULL)	/* probably an open error */
		return (0);

	lis_down(&fifo_sem);
	fifo_free(fdp->index);

	RD(q)->q_ptr = WR(q)->q_ptr = NULL;

	lis_up(&fifo_sem);

	qprocsoff(q);

	MODPUT();

	return 0;		/* success */
}

#if 0
/*
 *  wput
 */
static void
fifo_wput(queue_t *q, mblk_t *mp)
{
	fifo_dev_t *fdp = (fifo_dev_t *) q->q_ptr;

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_wput( 0x%08x, 0x%08x ) unit %d\n", MOD_NAME, q, mp, fdp->index);
#endif

	/* 
	 *  this is a driver - we should NAK IOCTLs we don't handle
	 *  (which means all of them)
	 */
	switch (mp->b_datap->db_type) {
	case M_IOCTL:
		mp->b_datap->db_type = M_IOCNAK;
		qreply(q, mp);
		break;
	default:
		putnext(q, mp);	/* this goes to our RD(q) */
		break;
	}
}

/*
 *  wsrv
 */
static void
fifo_wsrv(queue_t *q)
{
	fifo_dev_t *fdp = (fifo_dev_t *) q->q_ptr;
	mblk_t *mp;

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_wsrv( 0x%08x ) unit %d\n", MOD_NAME, q, fdp->index);
#endif

	while ((mp = getq(q))) {
		/* 
		 *  this is a driver - we should NAK IOCTLs we don't handle
		 *  (which means all of them)
		 */
		switch (mp->b_datap->db_type) {
		case M_IOCTL:
			mp->b_datap->db_type = M_IOCNAK;
			qreply(q, mp);
			break;
		default:
			putnext(q, mp);	/* this goes to our RD(q) */
			break;
		}
	}
}

/*
 *  rput
 */
static void
fifo_rput(queue_t *q, mblk_t *mp)
{
	fifo_dev_t *fdp = (fifo_dev_t *) q->q_ptr;

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_rput( 0x%08x, 0x%08x ) unit %d\n", MOD_NAME, q, mp, fdp->index);
#endif

	putnext(q, mp);
}

/*
 *  rsrv
 */
static void
fifo_rsrv(queue_t *q)
{
	fifo_dev_t *fdp = (fifo_dev_t *) q->q_ptr;
	mblk_t *mp;

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_rsrv( 0x%08x ) unit %d\n", MOD_NAME, q, fdp->index);
#endif

	while ((mp = getq(q))) {
		putnext(q, mp);
	}
}
#endif

/*
 *  initialization/termination
 *
 *  These can be invoked one of two ways:
 *   . for modules, it is easiest to call them from [init,cleanup]_module
 *   . otherwise, specify "initialize" and "terminate" in Config
 */

void
fifo_init(void)
{
	int i;
	int maxminors = FIFO_MAX_MINOR + 1;

#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_init() #%d\n", MOD_NAME, fifo_initialized);
#endif

	/* 
	 *  only initialize once (per load)...
	 */
	if (fifo_initialized++)
		return;

	/* 
	 *  initialize major/minor device data
	 */
	for (i = 0; i < fifo_units; i++) {
		fifo_dev_t *fdp = &(fifo_dev[i]);

		/* 
		 *  initialize per-unit data -
		 *
		 *  we treat the index 'i' as a device number of sorts here; we
		 *  let major() and minor() split it into parts.  This works as
		 *  long as the low x bits are minor number, and the upper y
		 *  bits are major.
		 */
		fdp->index = i;
		if ((i % maxminors) != 0) {
			fdp->flags = 0;
			list_add(&(fdp->list), free_fifos.prev);
			num_free_fifos++;
		} else {
			fdp->flags = FIFO_CLONABLE;
			INIT_LIST_HEAD(&(fdp->list));
			/* 
			 *  indicate that this is a FIFO pseudo device that allows
			 *  minors to reopen
			 */
			LIS_DEVFLAGS(fifo__0_majors[i / maxminors]) |=
			    LIS_MODFLG_FIFO | LIS_MODFLG_REOPEN;
		}
	}
#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_init() %d units available\n", MOD_NAME, num_free_fifos);
#endif

	lis_sem_init(&fifo_sem, 1);
}

void
fifo_term(void)
{
#ifdef FIFO_DEBUG
	cmn_err(CE_CONT, "%s_term() #%d\n", MOD_NAME, fifo_initialized);
#endif

	if (--fifo_initialized)
		return;

	SEM_DESTROY(&fifo_sem);
}

#ifdef LINUX
#if !defined __NO_VERSION__

/*
 *  Linux loadable module interface
 */

#ifdef KERNEL_2_5
int
fifo_mod_init(void)
#else
int
init_module(void)
#endif
{
	int ret = lis_register_strdev(FIFO__CMAJOR_0, &fifo_info,
				      fifo_units, MOD_NAME);

	if (ret < 0) {
		cmn_err(CE_CONT, "%s - unable to register driver.\n", MOD_NAME);
		return ret;
	}

	fifo_init();

	return 0;
}

#ifdef KERNEL_2_5
void
fifo_mod_cleanup(void)
#else
void
cleanup_module(void)
#endif
{
	fifo_term();

	if (lis_unregister_strdev(FIFO__CMAJOR_0) < 0)
		cmn_err(CE_CONT, "%s - unable to unregister driver.\n", MOD_NAME);
	return;
}

#ifdef KERNEL_2_5
module_init(fifo_mod_init);
module_exit(fifo_mod_cleanup);
#endif
#if defined(MODULE_LICENSE)
MODULE_LICENSE("GPL");
#endif
#if defined(MODULE_AUTHOR)
MODULE_AUTHOR("John Boyd <jaboydjr@protologos.net>");
#endif
#if defined(MODULE_DESCRIPTION)
MODULE_DESCRIPTION("STREAMS-based FIFO pseudo-driver");
#endif
#if defined(MODULE_ALIAS)
MODULE_ALIAS("streams-" __stringify(LIS_OBJNAME));
#endif

#endif				/* !defined __NO_VERSION__ */

#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(FIFO__CMAJOR_0));
MODULE_ALIAS("char-major-" __stringify(FIFO__CMAJOR_0) "-*");
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-" __stringify(FIFO__CMAJOR_0));
MODULE_ALIAS("char-major-" __stringify(FIFO__CMAJOR_0) "-0");
MODULE_ALIAS("/dev/fifo");
MODULE_ALIAS("/dev/fifo.0");
#endif
#endif
